sklearn/neighbors/base.py

```diff
@@ -776,15 +776,14 @@ def fit(self, X, y):
         else:
             self.outputs_2d_ = True
 
-        try:
-            check_classification_targets(y)
-        except ValueError as e:
+        if not is_valid_classification_targets(y):
+            # specialize the ValueError message with a more informative error.
             if issparse(y) and self.outputs_2d_:
                 raise ValueError("Sparse y is only supported for multilabel"
                                  " case (multioutput multiclass is not"
-                                 " supported)")
+                                 " supported). Got: %r" % y)
             else:
-                raise
+                raise ValueError("Unknown label type: %r" % y)
```

Comment:
The initial idea with @ogrisel was to change the ValueError's meassage for something more descriptive like this:

```py
  raise ValueError("Unknown classification label type. Got: %r" % y)
```

The main problem is that such message breaks this:

```sh
  $pytest sklearn/tests/test_common.py::test_non_meta_estimators
```
If needed I'll change the message in a separated PR, so that no common files are changed here.